plugins/inputs/solr/solr.go

```diff
@@ -0,0 +1,424 @@
+		CumulativeLookups   int     `json:"cumulative_lookups"`
+		Evictions           int     `json:"evictions"`
+		Hitratio            float64 `json:"hitratio,string"`
+		Hits                int     `json:"hits"`
+		Inserts             int     `json:"inserts"`
+		Lookups             int     `json:"lookups"`
+		Size                int     `json:"size"`
+		WarmupTime          int     `json:"warmupTime"`
+	} `json:"stats"`
+}
+
+// NewSolr return a new instance of Solr
+func NewSolr() *Solr {
+	return &Solr{
+		HTTPTimeout: internal.Duration{Duration: time.Second * 5},
+	}
+}
+
+// SampleConfig returns sample configuration for this plugin.
+func (s *Solr) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the plugin description.
+func (s *Solr) Description() string {
+	return "Read stats from one or more Solr servers or cores"
+}
+
+// Default settings
+func (s *Solr) setDefaults() ([][]string, int, error) {
+	var max int
+	cores := [][]string{}
+	if len(s.Cores) == 0 {
+		for n, server := range s.Servers {
+			adminCores := &AdminCores{}
+			if err := s.gatherData(fmt.Sprintf("%s%s", server, adminCoresPath), adminCores); err != nil {
+				return nil, 0, err
+			}
+			serverCores := []string{}
+			for coreName := range adminCores.Status {
+				serverCores = append(serverCores, coreName)
+			}
+			cores = append(cores, serverCores)
+			if len(cores[n]) > max {
+				max = len(cores[n])
+			}
+		}
+	} else {
+		cores = append(cores, s.Cores)
+		max = len(s.Cores)
+	}
+	return cores, max, nil
+}
+
+// Gather reads the stats from Solr and writes it to the
+// Accumulator.
+func (s *Solr) Gather(acc telegraf.Accumulator) error {
+	if s.client == nil {
+		client := s.createHTTPClient()
+		s.client = client
+	}
+	cores, max, err := s.setDefaults()
+	if err != nil {
+		return err
+	}
+
+	errChan := errchan.New(len(s.Servers) * max)
+	var wg sync.WaitGroup
+
+	for n, serv := range s.Servers {
+		for _, core := range cores[n] {
+			wg.Add(1)
+			go func(serv string, core string, acc telegraf.Accumulator) {
+				defer wg.Done()
+				if err := s.gatherCoreStats(fmt.Sprintf("%s/solr/%s"+mbeansPath, serv, core), core, acc); err != nil {
+					errChan.C <- err
+					return
+				}
+			}(serv, core, acc)
+		}
+	}
+	wg.Wait()
+	return errChan.Error()
+}
+
+func (s *Solr) createHTTPClient() *http.Client {
+	tr := &http.Transport{
+		ResponseHeaderTimeout: s.HTTPTimeout.Duration,
+	}
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   s.HTTPTimeout.Duration,
+	}
+
+	return client
+}
+
+func gatherCoreMetrics(mbeansJSON json.RawMessage, core, category string, acc telegraf.Accumulator) error {
+	var coreMetrics map[string]Core
+
+	measurementTime := time.Now()
+	if err := json.Unmarshal(mbeansJSON, &coreMetrics); err != nil {
+		return err
+	}
+	for name, metrics := range coreMetrics {
+		if strings.Contains(name, "@") {
+			continue
+		}
+		coreFields := map[string]interface{}{
+			"class_name":   metrics.Class,
```

Comment:
@phemmer I feel like those class names are a bit useless in tags but just in case I left them as fields. If there is any good reason to keep them as tags, yeah why not I can convert them to tags.